File: DSA/src/Questions/Array/BeautifulPairs.java
package Questions.Array;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/*

https://leetcode.com/problems/beautiful-pairs/description/
You are given two 0-indexed integer arrays nums1 and nums2 of the same length. A pair of indices (i,j) is called beautiful if|nums1[i] - nums1[j]| + |nums2[i] - nums2[j]| is the smallest amongst all possible indices pairs where i < j.

Return the beautiful pair. In the case that there are multiple beautiful pairs, return the lexicographically smallest pair.

Note that

|x| denotes the absolute value of x.
A pair of indices (i1, j1) is lexicographically smaller than (i2, j2) if i1 < i2 or i1 == i2 and j1 < j2.

Example 1:

Input: nums1 = [1,2,3,2,4], nums2 = [2,3,1,2,3]
Output: [0,3]
Explanation: Consider index 0 and index 3. The value of |nums1[i]-nums1[j]| + |nums2[i]-nums2[j]| is 1, which is the smallest value we can achieve.
Example 2:

Input: nums1 = [1,2,4,3,2,5], nums2 = [1,4,2,3,5,1]
Output: [1,4]
Explanation: Consider index 1 and index 4. The value of |nums1[i]-nums1[j]| + |nums2[i]-nums2[j]| is 1, which is the smallest value we can achieve.
 

Constraints:

2 <= nums1.length, nums2.length <= 105
nums1.length == nums2.length
0 <= nums1i <= nums1.length
0 <= nums2i <= nums2.length
Time Complexity: O(NlogN)
Space Complexity: O(N)
 */
public class BeautifulPairs {
    private List<int[]> points = new ArrayList<>();

    // Main method to find the beautiful pair of indices
    public int[] beautifulPair(int[] nums1, int[] nums2) {
        int n = nums1.length;
        // Map to group indices based on a unique key generated from nums1 and nums2
        // values
        Map<Long, List<Integer>> pl = new HashMap<>();

        // Populate the map with unique keys and corresponding indices
        for (int i = 0; i < n; ++i) {
            long z = f(nums1[i], nums2[i]);
            pl.computeIfAbsent(z, k -> new ArrayList<>()).add(i);
        }

        // Check if any key has more than one index
        for (int i = 0; i < n; ++i) {
            long z = f(nums1[i], nums2[i]);
            if (pl.get(z).size() > 1) {
                return new int[] { i, pl.get(z).get(1) };
            }
            points.add(new int[] { nums1[i], nums2[i], i });
        }

        // Sort points based on the first element (nums1 value)
        points.sort((a, b) -> a[0] - b[0]);

        // Perform a divide and conquer search to find the beautiful pair
        int[] ans = dfs(0, points.size() - 1);
        return new int[] { ans[1], ans[2] };
    }

    // Helper method to generate a unique key from nums1 and nums2 values
    private long f(int x, int y) {
        return x * 100000L + y;
    }

    // Calculate Manhattan distance between two points
    private int dist(int x1, int y1, int x2, int y2) {
        return Math.abs(x1 - x2) + Math.abs(y1 - y2);
    }

    // Divide and conquer method to find the beautiful pair
    private int[] dfs(int l, int r) {
        // Base case: If only one element, return a large distance (infinity)
        if (l >= r) {
            return new int[] { 1 << 30, -1, -1 };
        }

        int m = (l + r) >> 1;
        int x = points.get(m)[0];

        // Recursive calls for left and right halves
        int[] t1 = dfs(l, m);
        int[] t2 = dfs(m + 1, r);

        // Choose the pair with the smaller distance
        if (t1[0] > t2[0] || (t1[0] == t2[0] && (t1[1] > t2[1] || (t1[1] == t2[1] && t1[2] > t2[2])))) {
            t1 = t2;
        }

        List<int[]> t = new ArrayList<>();

        // Collect points that are within the current best distance
        for (int i = l; i <= r; ++i) {
            if (Math.abs(points.get(i)[0] - x) <= t1[0]) {
                t.add(points.get(i));
            }
        }

        // Sort the collected points based on the second element (nums2 value)
        t.sort((a, b) -> a[1] - b[1]);

        // Check pairs within the collected points to find the smallest distance
        for (int i = 0; i < t.size(); ++i) {
            for (int j = i + 1; j < t.size(); ++j) {
                if (t.get(j)[1] - t.get(i)[1] > t1[0]) {
                    break;
                }
                int pi = Math.min(t.get(i)[2], t.get(j)[2]);
                int pj = Math.max(t.get(i)[2], t.get(j)[2]);
                int d = dist(t.get(i)[0], t.get(i)[1], t.get(j)[0], t.get(j)[1]);
                if (d < t1[0] || (d == t1[0] && (pi < t1[1] || (pi == t1[1] && pj < t1[2])))) {
                    t1 = new int[] { d, pi, pj };
                }
            }
        }
        return t1;
    }
    public static void main(String[] args) {
        BeautifulPairs beautifulPairs = new BeautifulPairs();
        int[] nums1 = {1, 2, 3, 2, 4};
        int[] nums2 = {2, 3, 1, 2, 3};
        int[] result = beautifulPairs.beautifulPair(nums1, nums2);
        System.out.println(result[0] + " " + result[1]);
    }
}

File: DSA/Python/Questions/Array/BeautifulPairs.py
"""
# Question: Beautiful Pairs
# Link: https://leetcode.com/problems/beautiful-pairs/

# Given two arrays nums1 and nums2 of equal length n, find the pair of indices (i,j)
# such that the Manhattan distance between the elements at those indices is minimized.
# The Manhattan distance between two points (x1,y1) and (x2,y2) is |x1-x2| + |y1-y2|.

# Time Complexity: O(n log n) where n is the length of input arrays
# Space Complexity: O(n) for storing points list

# Algorithm:
# 1. Convert array elements into points with coordinates (nums1[i], nums2[i])
# 2. Use divide and conquer approach to find minimum Manhattan distance:
#    - Divide points into left and right halves
#    - Recursively find minimum distances in both halves
#    - Combine results by checking points near the dividing line
# 3. For points near dividing line:
#    - Sort by y-coordinate
#    - Compare each point with nearby points
#    - Update minimum distance if smaller distance found

# Key Components:
# - dist(): Calculates Manhattan distance between two points
# - dfs(): Implements divide and conquer strategy
# - beautiful_pair(): Main function that processes input arrays

# Data Structures:
# - self.points: List of tuples storing (x, y, index) for each point
# - Return format: [distance, index1, index2] where indices form the beautiful pair
"""

# Import List type from typing module
from typing import List


# Define BeautifulPairs class
class BeautifulPairs:
    # Initialize class with empty points list
    def __init__(self):
        # List to store points
        self.points = []

    # Calculate Manhattan distance between two points
    def dist(self, x1: int, y1: int, x2: int, y2: int) -> int:
        # Return absolute difference of x coordinates plus absolute difference of y coordinates
        return abs(x1 - x2) + abs(y1 - y2)

    # Recursive divide and conquer function
    def dfs(self, l: int, r: int) -> List[int]:
        # Base case: if left >= right, return max value and invalid indices
        if l >= r:
            return [1 << 30, -1, -1]

        # Calculate middle point
        m = (l + r) >> 1
        # Get x coordinate of middle point
        x = self.points[m][0]

        # Recursively process left half
        t1 = self.dfs(l, m)
        # Recursively process right half
        t2 = self.dfs(m + 1, r)

        # Compare results from both halves and take the better one
        if t1[0] > t2[0] or (
            t1[0] == t2[0] and (t1[1] > t2[1] or (t1[1] == t2[1] and t1[2] > t2[2]))
        ):
            t1 = t2

        # Create list for points near dividing line
        t = []
        # Collect points that are within current minimum distance of dividing line
        for i in range(l, r + 1):
            if abs(self.points[i][0] - x) <= t1[0]:
                t.append(self.points[i])

        # Sort points by y coordinate
        t.sort(key=lambda x: x[1])

        # Compare each point with nearby points
        for i in range(len(t)):
            for j in range(i + 1, len(t)):
                # Break if y distance exceeds current minimum
                if t[j][1] - t[i][1] > t1[0]:
                    break
                # Get indices in sorted order
                pi = min(t[i][2], t[j][2])
                pj = max(t[i][2], t[j][2])
                # Calculate distance between points
                d = self.dist(t[i][0], t[i][1], t[j][0], t[j][1])
                # Update result if better distance found
                if d < t1[0] or (
                    d == t1[0] and (pi < t1[1] or (pi == t1[1] and pj < t1[2]))
                ):
                    t1 = [d, pi, pj]
        # Return best result
        return t1

    # Main function to find beautiful pair
    def beautiful_pair(self, nums1: List[int], nums2: List[int]) -> List[int]:
        # Get length of input arrays
        n = len(nums1)
        # Reset points list
        self.points = []
        # Create points from input arrays
        for i in range(n):
            self.points.append([nums1[i], nums2[i], i])
        # Sort points by x coordinate
        self.points.sort()
        # Get result from divide and conquer
        result = self.dfs(0, n - 1)
        # Return indices of beautiful pair
        return [result[1], result[2]]


# Main function for testing
def main():
    # Create instance of BeautifulPairs
    beautiful_pairs = BeautifulPairs()
    # Test input arrays
    nums1 = [1, 2, 3, 2, 4]
    nums2 = [2, 3, 1, 2, 3]
    # Get result
    result = beautiful_pairs.beautiful_pair(nums1, nums2)
    # Print result
    print(f"{result[0]} {result[1]}")


# Execute main function if script is run directly
if __name__ == "__main__":
    main()
